indra/llcommon/llcallbacklist.h

```diff
@@ -27,53 +27,248 @@
+    // Return true iff the handle corresponds to a live timer.
+    bool cancel(const handle_t& timer);
+    // If we're canceling a non-const handle_t, also clear it so we need not
+    // cancel again.
+    bool cancel(handle_t& timer);
+
+    // Store a handle_t returned by doAtTime(), doAfterInterval() or
+    // doPeriodically() in a temp_handle_t to cancel() automatically on
+    // destruction of the temp_handle_t.
+    class temp_handle_t
+    {
+    public:
+        temp_handle_t() {}
+        temp_handle_t(const handle_t& hdl): mHandle(hdl) {}
+        temp_handle_t(const temp_handle_t&) = delete;
+        temp_handle_t(temp_handle_t&&) = default;
+        temp_handle_t& operator=(const handle_t& hdl)
+        {
+            // initializing a new temp_handle_t, then swapping it into *this,
+            // takes care of destroying any previous mHandle
+            temp_handle_t replacement(hdl);
+            swap(replacement);
+            return *this;
+        }
+        temp_handle_t& operator=(const temp_handle_t&) = delete;
+        temp_handle_t& operator=(temp_handle_t&&) = default;
+        ~temp_handle_t()
+        {
+            cancel();
+        }
+
+        // temp_handle_t should be usable wherever handle_t is
+        operator handle_t() const { return mHandle; }
+        // If we're dealing with a non-const temp_handle_t, pass a reference
+        // to our handle_t member (e.g. to LLLater::cancel()).
+        operator handle_t&() { return mHandle; }
+
+        // For those in the know, provide a cancel() method of our own that
+        // avoids LLLater::instance() lookup when mHandle isn't live.
+        bool cancel()
+        {
+            if (! mHandle)
+            {
+                return false;
+            }
+            else
+            {
+                return LLLater::instance().cancel(mHandle);
+            }
+        }
+
+        void swap(temp_handle_t& other) noexcept
+        {
+            std::swap(this->mHandle, other.mHandle);
+        }
+
+    private:
+        handle_t mHandle;
+    };
+
+private:
+    bool tick();
+
+    // NOTE: We don't lock our data members because it doesn't make sense to
+    // register cross-thread callbacks. If we start wanting to use them on
+    // threads other than the main thread, it would make more sense to make
+    // our data members thread_local than to lock them.
+
+    // the heap aka priority queue
+    queue_t mQueue;
+    // handles we've returned that haven't yet canceled
+    using HandleMap = std::unordered_map<
+        token_t,
+        std::pair<queue_t::handle_type, LLDate::timestamp>>;
+    HandleMap mHandles;
+    token_t mToken{ 0 };
+    // While mQueue is non-empty, register for regular callbacks.
+    LLCallbackList::temp_handle_t mLive;
+
+    struct Periodic;
+
+    // internal implementation for doAtTime()
+    HandleMap::iterator doAtTime1(LLDate::timestamp time);
+    handle_t doAtTime2(nullary_func_t callable, HandleMap::iterator iter);
```

Comment:
I'm not sure why we need doAtTime1/2 here, when the compiler is perfectly capable to distinguish them by their arguments. 

Also
- `dotAtTime1` would be in lieu of my suggestion of using `schedule... `
- `doAtTime2 suddenly decides to be totally different, while all previous methods prefixed with doAt accepted some sort of interval or time, now there is an iterator passed.
-

Reply:
Those internal-only names are because, to implement `doPeriodically()`, I found it necessary to split the original implementation of `doAtTime()` into two other methods. They literally mean "first steps of `doAtTime()`," "second steps of `doAtTime()`."

But it looks like I'm going to have to refactor all that anyway.